File: bst/binary_search_tree.py
import sys

# БИНАРНОЕ ДЕРЕВО ПОИСКА

class Node:
    def __init__(self, value=None):
        self.value = value
        self.parent = None
        self.left_child = None
        self.right_child = None

    def untie(self):
        if self.parent:
            self.parent = None

        self.left_child = None
        self.right_child = None


class Tree:
    root = None

    def _insert(self, current_node, node):
        if node.value <= current_node.value:
            if not current_node.left_child:
                current_node.left_child = node
                node.parent = current_node
                return
            self._insert(current_node.left_child, node)
        else:
            if not current_node.right_child:
                current_node.right_child = node
                node.parent = current_node
                return
            self._insert(current_node.right_child, node)

    def insert(self, value):
        node = Node(value)

        if not self.root:
            self.root = node
            return

        self._insert(self.root, node)

    def preorder(self):
        return self._preorder(self.root)

    def _preorder(self, node: Node):
        if node is None:
            return []

        result = [node]
        result += self._preorder(node.left_child)
        result += self._preorder(node.right_child)
        return result

    def inorder(self):
        return self._inorder(self.root)

    def _inorder(self, node: Node):
        if node is None:
            return []

        result = self._inorder(node.left_child)
        result.append(node)
        result += self._inorder(node.right_child)
        return result

    def postorder(self):
        return self._postorder(self.root)

    def _postorder(self, node: Node):
        if node is None:
            return []

        result = self._postorder(node.left_child)
        result += self._postorder(node.right_child)
        result.append(node)
        return result

    def minimum(self):
        return self._minimum(self.root)

    def _minimum(self, current_node: Node):
        if current_node is None:
            return

        return self._minimum(current_node.left_child) if current_node.left_child else current_node

    def maximum(self):
        return self._maximum(self.root)

    def _maximum(self, current_node: Node):
        if current_node is None:
            return

        return self._maximum(current_node.right_child) if current_node.right_child else current_node

    def find(self, value) -> Node:
        return self._find(value, self.root)

    def _find(self, value, current_node: Node) -> Node:
        if current_node is None:
            return None
        if current_node.value == value:
            return current_node

        if current_node.value < value:
            if current_node.right_child:
                return self._find(value, current_node.right_child)
        elif current_node.left_child:
            return self._find(value, current_node.left_child)

    def next_to(self, value):
        return self._next_to(value, self.root)

    def _next_to(self, value, current_node: Node, successor: Node = None):
        if current_node is None:
            return successor

        if value < current_node.value:
            return self._next_to(value, current_node.left_child, current_node)
        else:
            return self._next_to(value, current_node.right_child, successor)

    def prev_to(self, value):
        return self._prev_to(value, self.root)

    def _prev_to(self, value, current_node: Node, successor: Node = None):
        if current_node is None:
            return successor

        if value <= current_node.value:
            return self._prev_to(value, current_node.left_child, successor)
        else:
            return self._prev_to(value, current_node.right_child, current_node)

    def _delete(self, node):
        if node is None:
            return
        if not node.left_child and not node.right_child:
            if node is self.root:
                self.root = None
            elif node.parent.left_child is node:
                node.parent.left_child = None
            elif node.parent.right_child is node:
                node.parent.right_child = None
            node.untie()
            return node
        elif node.left_child and node.right_child:
            next_node = self.next_to(node.value)
            self._delete(next_node)
            node.value = next_node.value
            return
        elif node.left_child:
            if node.parent:
                if node.parent.left_child is node:
                    node.parent.left_child = node.left_child
                elif node.parent.right_child is node:
                    node.parent.right_child = node.left_child
                node.left_child.parent = node.parent
            if node is self.root:
                self.root = node.left_child
            node.untie()
        elif node.right_child:
            if node.parent:
                if node.parent.left_child is node:
                    node.parent.left_child = node.right_child
                elif node.parent.right_child is node:
                    node.parent.right_child = node.right_child
                node.right_child.parent = node.parent
            if node is self.root:
                self.root = node.right_child
            node.untie()
        return node

    def delete(self, value):
        node = self.find(value)
        if node is None:
            return

        self._delete(node)


"""
Эта программа умеет отвечать на запросы:

insert x - вставка в дерево элемента x
delete x - удаление из дерева элемента x
exists x - проверка на наличие элемента x (true, если элемент в дереве, иначе false)
next x - вывод следующего элемента после x (none, если такового нет)
prev x - вывод предыдущего элемента перед x (none, если такового нет)
"""

t = Tree()

for i in sys.stdin:
    operation, number = i.split()
    number = int(number)
    if operation == 'insert':
        node = t.find(number)
        if node:
            continue
        t.insert(number)
    elif operation == 'exists':
        node = t.find(number)
        print('true' if node else 'false')
    elif operation == 'next':
        node = t.next_to(number)
        print(node.value if node else 'none')
    elif operation == 'prev':
        node = t.prev_to(number)
        print(node.value if node else 'none')
    elif operation == 'delete':
        t.delete(number)

File: bst/binary_search_tree.cpp
#include <iostream>
#include <vector>

// БИНАРНОЕ ДЕРЕВО ПОИСКА

template<typename T>
struct Node {
    T value;
    Node<T>* parent = nullptr;
    Node<T>* left_child = nullptr;
    Node<T>* right_child = nullptr;
    int children = 1;

    void untie() {
        parent = nullptr;
        left_child = nullptr;
        right_child = nullptr;
    }
    void rechildren() {
        int new_children = 1;
        if (left_child) {
            new_children += left_child->children;
        }
        if (right_child) {
            new_children += right_child->children;
        }
        if (parent && children != new_children) {
            children = new_children;
            parent->rechildren();
        }
        children = new_children;
    }

};

template<typename T>
struct Tree {
    Node<T>* root = nullptr;

    void _insert(Node<T>* current_node, Node<T>* node) {
        if (node->value < current_node->value) {
            if (!current_node->left_child) {
                current_node->left_child = node;
                node->parent = current_node;
            } else {
                _insert(current_node->left_child, node);
            }
        } else if (node->value > current_node->value) {
            if (!current_node->right_child) {
                current_node->right_child = node;
                node->parent = current_node;
            } else {
                _insert(current_node->right_child, node);
            }
        } else {
            return;
        }
        current_node->rechildren();
    }

    Node<T>* next_to(T value) {
        return next_to(value, root);
    }

    Node<T>* next_to(T value, Node<T>* current_node, Node<T>* successor = nullptr) {
        if (!current_node) {
            return successor;
        }
        if (value < current_node->value) {
            return next_to(value, current_node->left_child, current_node);
        } else {
            return next_to(value, current_node->right_child, successor);
        }
    }

    Node<T>* prev_to(T value) {
        return prev_to(value, root);
    }

    Node<T>* prev_to(T value, Node<T>* current_node, Node<T>* successor = nullptr) {
        if (!current_node) {
            return successor;
        }
        if (value <= current_node->value) {
            return prev_to(value, current_node->left_child, successor);
        } else {
            return prev_to(value, current_node->right_child, current_node);
        }
    }

    void insert(const T& value) {
        insert(new Node<T>{value});
    }

    void insert(Node<T>* node) {
        if (!node) {
            return;
        }
        if (!root) {
            root = node;
            return;
        }
        _insert(root, node);
    }

    Node<T>* find(T value) {
        return find(value, root);
    }

    Node<T>* find(T value, Node<T>* current_node) {
        if (!current_node) {
            return nullptr;
        }
        if (current_node->value == value) {
            return current_node;
        } else if (current_node->value < value) {
            if (current_node->right_child) {
                return find(value, current_node->right_child);
            }
        } else {
            if (current_node->left_child) {
                return find(value, current_node->left_child);
            }
        }
        return nullptr;
    }

    Node<T>* remove(T value) {
        Node<T>* place = find(value, root);
        if (!place) {
            return nullptr;
        }
        return _remove(place);
    }

    Node<T>* _remove(Node<T>* node) {
        if (!node) {
            return nullptr;
        }
        if (!node->left_child && !node->right_child) {
            if (node == root) {
                root = nullptr;
            } else if (node->parent->left_child == node) {
                node->parent->left_child = nullptr;
                node->parent->rechildren();
            } else if (node->parent->right_child == node) {
                node->parent->right_child = nullptr;
                node->parent->rechildren();
            }
            node->untie();
        } else if (node->left_child && node->right_child) {
            Node<T>* next_node = next_to(node->value, root);
            Node<T>* result = _remove(next_node);
            node->value = next_node->value;
            return result;
        } else if (node->left_child) {
            if (node->parent) {
                if (node->parent->left_child == node) {
                    node->parent->left_child = node->left_child;
                } else if (node->parent->right_child == node) {
                    node->parent->right_child = node->left_child;
                }
                node->left_child->parent = node->parent;
                node->parent->rechildren();
            }
            if (node == root) {
                root = node->left_child;
            }
            node->untie();
        } else if (node->right_child) {
            if (node->parent) {
                if (node->parent->left_child == node) {
                    node->parent->left_child = node->right_child;
                } else if (node->parent->right_child == node) {
                    node->parent->right_child = node->right_child;
                }
                node->right_child->parent = node->parent;
                node->parent->rechildren();
            }
            if (node == root) {
                root = node->right_child;
            }
            node->untie();
        }

        return node;
    }

    std::vector<T> inorder() {
        return inorder(root);
    }

    std::vector<T> inorder(Node<T>* current_node) {
        if (!current_node) {
            return {};
        }
        auto result = inorder(current_node->left_child);
        result.push_back(current_node->value);
        auto result2 = inorder(current_node->right_child);
        result.insert(result.end(), result2.begin(), result2.end());

        return result;
    }

    std::vector<T> preorder() {
        return preorder(root);
    }

    std::vector<T> preorder(Node<T>* current_node) {
        if (!current_node) {
            return {};
        }
        std::vector<T> result = {current_node->value};
        auto result2 = preorder(current_node->left_child);
        result.insert(result.end(), result2.begin(), result2.end());
        auto result3 = preorder(current_node->right_child);
        result.insert(result.end(), result3.begin(), result3.end());

        return result;
    }

    std::vector<T> postorder() {
        return postorder(root);
    }

    std::vector<T> postorder(Node<T>* current_node) {
        if (!current_node) {
            return {};
        }
        auto result = postorder(current_node->left_child);
        auto result2 = postorder(current_node->right_child);
        result.insert(result.end(), result2.begin(), result2.end());
        result.push_back(current_node->value);

        return result;
    }

    Node<T>* find_k_element(size_t k, Node<T>* current_node) {
        if (!current_node) {
            return nullptr;
        }
        if (current_node->left_child) {
            if (current_node->left_child->children > k) {
                return find_k_element(k, current_node->left_child);
            }
            k -= current_node->left_child->children;
        }
        if (k == 0) {
            return current_node;
        } else {
            k--;
        }
        return find_k_element(k, current_node->right_child);
    }

};


/**
Эта программа умеет отвечать на запросы:

insert x - вставка в дерево элемента x
delete x - удаление из дерева элемента x
exists x - проверка на наличие элемента x (true, если элемент в дереве, иначе false)
next x - вывод следующего элемента после x (none, если такового нет)
prev x - вывод предыдущего элемента перед x (none, если такового нет)
index k - вывод элемента дерева под индексом k (индексация с 0)
 */

int main() {
    int x;
    Tree<int> t;

    std::string operation;
    while (std::cin >> operation >> x) {
        if (operation == "insert") {
            t.insert(x);
        } else if (operation == "exists") {
            std::cout << (t.find(x) ? "true" : "false") << std::endl;
        } else if (operation == "delete") {
            t.remove(x);
        } else if (operation == "next") {
            auto node = t.next_to(x);
            if (node) {
                std::cout << node->value << std::endl;
            } else {
                std::cout << "none" << std::endl;
            }
        } else if (operation == "prev") {
            auto node = t.prev_to(x);
            if (node) {
                std::cout << node->value << std::endl;
            } else {
                std::cout << "none" << std::endl;
            }
        } else if (operation == "index") {
            auto node = t.find_k_element(x, t.root);
            if (node) {
                std::cout << node->value << std::endl;
            } else {
                std::cout << "none" << std::endl;
            }
        }
    }

    return 0;
}
